src/librustdoc/clean/types.rs

```diff
@@ -48,7 +49,82 @@ use self::ItemKind::*;
 use self::SelfTy::*;
 use self::Type::*;
 
-thread_local!(crate static MAX_DEF_IDX: RefCell<FxHashMap<CrateNum, DefIndex>> = Default::default());
+crate type DefIdSet = FxHashSet<DefId>;
+crate type RealDefId = rustc_hir::def_id::DefId;
+
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Copy)]
+crate enum DefId {
+    Real(RealDefId),
+    Fake(usize),
+}
+
+impl DefId {
+    crate fn new_fake() -> Self {
+        thread_local!(static FAKE_DEF_ID_COUNTER: Cell<usize> = Cell::new(0));
+        let id = FAKE_DEF_ID_COUNTER.with(|id| {
+            let tmp = id.get();
+            id.set(tmp + 1);
+            tmp
+        });
+        Self::Fake(id)
+    }
+
+    crate fn new_real(id: rustc_hir::def_id::DefId) -> Self {
+        Self::Real(id)
+    }
+
+    #[inline]
+    crate fn is_local(self) -> bool {
+        match self {
+            DefId::Real(id) => id.is_local(),
+            DefId::Fake(_) => false,
```

Comment:
This seems wrong. Right now, `is_local` checks the crate number and works even with fake IDs: https://doc.rust-lang.org/nightly/nightly-rustc/src/rustc_span/def_id.rs.html#253-255

If you look at `next_fake_id` currently, it takes the crate number. I think we should continue to do that and base `is_local` of the crate. `Fake(usize, CrateNum)` is probably the easiest way to do that.